### PR TITLE
Skip crafting submenus if no options available

### LINKED ISSUE #23314: Hide or skip crafting submenus if no options available
Game version:
b87c6d2

Minor UX thing: would be a nice simplification to hide or skip over crafting submenus if the user has no recipes in it yet.

screen shot 2018-03-28 at 8 36 21 pm

### PR DESCRIPTION
**Pull request by PatrikStridvall**
After reading the discussion concerning issue #23314, I did the following patch.

EDIT: If accepted this closes #23314.

### CODE REVIEW

**Comment by illi-kun** [comment]
@PatrikStridvall please use keywords from [this article](https://help.github.com/articles/closing-issues-using-keywords/) to close the fixed issue automatically. In short, adding a phrase "closes #23314" in the initial message of PR doing the trick.

**Comment by PatrikStridvall** [comment]
@illi-kun OK. I have edited the initial message. Hopefully this is what you meant.

**Comment by illi-kun** [comment]
Exactly! Thanks, this reduces the amount of brainless work for the repository's maintainers.

File: src/recipe_dictionary.cpp [hunk 1]
```diff
@@ -121,6 +121,13 @@ std::vector<const recipe *> recipe_subset::search( const std::string &txt,
     return res;
 }
 
+bool recipe_subset::empty_category( const std::string &cat,
+                                    const std::string &subcat ) const
+{
+    // @todo: Make more efficient
+    return in_category( cat, subcat ).size() == 0;
```
Comment (illi-kun) [inline src/recipe_dictionary.cpp]: why not just ```.empty()```?

Comment (PatrikStridvall) [reply src/recipe_dictionary.cpp]: If it is the todo comment you are referring what I meant it than in_category builds a vector that are not really needed however I was tired a wasn't in the mood to copy and adapt the code of in_category.

If you are referring to that .empty() for vectors is marginally more efficient that .size() == 0, then you are right as well. C++ is not my main language, I prefer Python, so I do not always select the most optimal way to do it when programming in C++. In addition since most of what I do professionally has database access as the dominant source of slowness. I usually care more about readability than speed.

In any case, since you have mentioned it, I will try to implement what I meant by the todo comment, later. However, since needlessly calculating and sizing a small vector is very fast anyway, I don't consider it a priority.

Premature optimization is the root to all evil. :-)

In short, noted, but I don't consider it a show stopper for merging.

Comment (PatrikStridvall) [reply src/recipe_dictionary.cpp]: @illi-kun I have made the method you complained about more efficient now and resubmitted it rebase onto the latest master. Especially see commit 15b6313. This others are the same, I just split them in separate commits.

Comment (PatrikStridvall) [reply src/recipe_dictionary.cpp]: Perhaps I should mention that I experimented with making the disabled subtabs italic in addition to dark gray. However, I only got it to work on the non-tiles version so I put i on ice. I will revisit the issue at a later time. Until then dark gray will have to suffice.

**Comment by narc0tiq** [comment]
> @narc0tiq, mind if I ping you about gorgon? Can you please check if PR Builder is hung?

I don't mind -- quite the opposite, I appreciate the notification. Turns out yes, the Jenkins was totally hung from a "no space left on device". I cleared some space and rebooted it, but it's not come back up yet. I've poked someone who can control the virtual machine and hopefully get it force booted.

**Comment by ZhilkinSerg** [comment]
> I don't mind -- quite the opposite, I appreciate the notification. Turns out yes, the Jenkins was totally hung from a "no space left on device". I cleared some space and rebooted it, but it's not come back up yet. I've poked someone who can control the virtual machine and hopefully get it force booted.

Thanks!

File: src/crafting_gui.cpp [hunk 2]
```diff
@@ -226,7 +226,7 @@ const recipe *select_crafting_recipe( int &batch_size )
 
             TAB_MODE m = ( batch ) ? BATCH : ( filterstring.empty() ) ? NORMAL : FILTERED;
             draw_recipe_tabs( w_head, tab.cur(), m );
-            draw_recipe_subtabs( w_subhead, tab.cur(), subtab.cur(), m );
+            draw_recipe_subtabs( w_subhead, tab.cur(), subtab.cur(), available_recipes );
```
Comment (BevapDin) [inline src/crafting_gui.cpp]: Why did you remove the argument `m` from the function call? As this is the only place that calls the function, it's never called with an explicit value for that `TAB_MODE mode` parameter (the default value will be used instead). You might as well remove that parameter from the function as it's now useless.

**Comment by PatrikStridvall** [comment]
@BevapDin You are absolutely correct. Programming after midnight tend to lead to stupid errors and since it was an argument with a default value, the compiler was not much help here. Luckily we have code reviewers like you. :1st_place_medal:

Anyway, I have a fixup in #23363.
